### PR TITLE
DDF-4642 Update FeatureMetacardType to handle simpleTypes and simpleContent to arbitrary depths

### PR DESCRIPTION
**Pull request by jrnorth**
#### What does this PR do?

Previously, when evaluating a feature property whose type is a simpleType, FeatureMetacardType would only go one level deep to determine its ‘base’ type. The ‘base’ type is used to determine what
AttributeType to use to map to the feature property, e.g. string, date, or integer. If the ‘base’ type is a type we can map to an AttributeType then FeatureMetacardType adds the feature property to itself as an attribute descriptor. Only going one level deep meant that feature properties whose type was a ‘base’ schema type such as xsd:string or xsd:int would be mapped and added as attribute descriptor, but feature properties whose type was a custom-defined simpleType extending a ‘base’ type would not be because we’d look at its QName, see it’s not a ‘base’ type, and pass over it. As a result, feature properties whose type was a simpleType were not queryable at all.

This commit changes how FeatureMetacardType handles simpleTypes. Now it will attempt to find a simpleType’s root type by following <list> and <restriction> definitions until a ‘base’ schema type is reached.

This commit also updates FeatureMetacardType to handle simpleContent inside complexTypes. Now simpleContent `<restriction>`s and `<extension>`s are handled in virtually the same way simpleTypes are handled as described above.

The upshot is that WFS sources can now query feature properties whose type is a simpleType or a complexType with simpleContent and that type’s root type is a ‘base’ schema type that DDF can understand.

The three FeatureMetacardTypeTest classes in their respective WFS 1.0.0, 1.1.0, and 2.0.0 modules have been consolidated into one FeatureMetacardTypeTest that lives in the spatial-wfs-common module. FeatureMetacardType lives in the spatial-wfs-common module so the unit tests for it in the other three modules did not count for unit test coverage against FeatureMetacardType. Moreover, the tests in the three classes were virtually identical so consolidating them into one made sense.

Finally, the XmlSchema object used in the FeatureMetacardType unit tests is now obtained by reading a real schema file instead of being constructed manually. The schema object trees constructed in the tests did not accurately model the schema object trees seen at runtime when a real schema file is parsed. Now that the unit tests obtain the XmlSchema object in the same way the runtime code does, they will be identical.

#### Who is reviewing it? 
@glenhein 
@leo-sakh 
@Corey-Collins 

#### Select relevant component teams: 
@codice/io 
@codice/ogc 

#### Ask 2 committers to review/merge the PR and tag them here.
@bdeining
@troymohl

#### How should this be tested?
1. Create a WFS 1.1.0 source with the URL http://services.azgs.az.gov/ArcGIS/services/OneGeology/AZGS_HIGS_Geology_WFS/MapServer/WFSServer
a. Disable the CN check and allow redirects. Leave everything else as-is.
2. In the Karaf console, run `log:set info org.apache.cxf` and `log:set warn org.apache.camel`.
3. Open Intrigue.
4. Run an anyText * query on the source. It may take a while to complete.
5. Check DDF's logs for an `Outbound Message` to http://services.azgs.az.gov/ArcGIS/services/OneGeology/AZGS_HIGS_Geology_WFS/MapServer/WFSServer. The request body should be a `GetFeature` XML request.
6. Open http://services.azgs.az.gov/ArcGIS/services/OneGeology/AZGS_HIGS_Geology_WFS/MapServer/WFSServer?SERVICE=WFS&REQUEST=DescribeFeatureType.
7. Check the GetFeature XML in the console and verify that all of the `element`s inside `GeologicUnitViewType` from step 6 _except_ `shape`, `OBJECTID`, `shape_Length`, and `shape_Area` are present in the query.
a. These elements are all custom-defined `simpleType`s that would not have been included in the query without this change.

#### What are the relevant tickets?
Fixes: #4642 

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Threat Dragon models
- [x] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: catalog/spatial/wfs/spatial-wfs-common/src/main/java/org/codice/ddf/spatial/ogc/wfs/catalog/common/FeatureMetacardType.java [hunk 1]
```diff
@@ -237,28 +242,95 @@ private void processXmlSchemaParticle(XmlSchemaParticle particle) {
...
+    if (baseTypeName != null) {
+      mapSchemaElement(parentElement, baseTypeName);
+    } else {
+      LOGGER.debug("Could not find the base type for simple content: {}", simpleContentTypeName);
+    }
+  }
+
+  private void processSimpleType(final XmlSchemaElement xmlSchemaElement) {
+    final QName baseTypeName =
+        xmlSchemaElement.getSchemaType() != null
```
Comment (glenhein) [inline catalog/spatial/wfs/spatial-wfs-common/src/main/java/org/codice/ddf/spatial/ogc/wfs/catalog/common/FeatureMetacardType.java]: Should this be `xmlSchemaElement.getSchemaType() intanceof XmlSchemaSimpleType`?

Comment (jrnorth) [reply catalog/spatial/wfs/spatial-wfs-common/src/main/java/org/codice/ddf/spatial/ogc/wfs/catalog/common/FeatureMetacardType.java]: Actually, I can remove this ternary. `processSimpleType()` is only called in two places, each of which is inside an `if` block that checks that the schema type of the element passed to this function is an `XmlSchemaSimpleType`.

File: catalog/spatial/wfs/spatial-wfs-common/src/test/java/org/codice/ddf/spatial/ogc/wfs/catalog/common/FeatureMetacardTypeTest.java [hunk 2]
```diff
@@ -0,0 +1,572 @@
...
+            new ValidationAttributes(),
+            new MediaAttributes())
+        .map(MetacardType::getAttributeDescriptors)
+        .flatMap(Set::stream)
+        .collect(Collectors.toSet());
+  }
+
+  private void loadSchema(final String schemaFile) throws IOException {
+    final XmlSchemaCollection schemaCollection = new XmlSchemaCollection();
+    try (final InputStream schemaStream = new FileInputStream("src/test/resources/" + schemaFile)) {
```
Comment (bakejeyner) [inline catalog/spatial/wfs/spatial-wfs-common/src/test/java/org/codice/ddf/spatial/ogc/wfs/catalog/common/FeatureMetacardTypeTest.java]: ✏️ Alternatively load the file stream using the classloader
```
try (final InputStream schemaStream = getClass().getClassLoader().getResourceAsStream(schemaFile)) {
```

**Comment by jrnorth** [comment]
build now

**Comment by jrnorth** [comment]
build now

**Comment by willwill96** [comment]
Hero Started

**Comment by willwill96** [comment]
### Hero Successful 
1. Setup WFS Source as specified in instructions
2. Turned up logging as specified and didn't notice anything suspicious during hero process
3. Ran an anyText * search on the WFS Source
4. Found the `Outbound Message` to the URL posted in the logs and verified it was a `GetFeature` request
5. Opened the URL posted with URL parameters
6. Compared the `GetFeature` XML response to the response in step 5 and found them to be identical with the exception of the 4 elements instructed to ignore in the hero instructions 

LGTM

**Pull request merged**